xc/device.ml

```diff
@@ -1631,6 +1631,50 @@ let vgpu_args_of_nvidia domid vcpus (vgpu:Xenops_interface.Vgpu.nvidia) =
 	then ["--resume=" ^ resume_file]
 	else []
 
+let write_vgpu_data ~xs domid devid keys =
+	let path = xenops_vgpu_path domid devid in
+	xs.Xs.writev path keys
+
+let mxgpu_device_in_use ~xs physical_function =
+	(* Check if there is a /xenops/domain/<x>/device/vgpu/<y>/pf xenstore node
+	 * that is equal to the given physical_function. *)
+	let root = Device_common.xenops_domain_path in
+	try
+		xs.Xs.directory root
+		|> List.map (fun domid ->
+				let path = Printf.sprintf "%s/%s/device/vgpu" root domid in
+				try
+					List.map (fun x -> path ^ "/" ^ x) (xs.Xs.directory path)
+				with Xs_protocol.Enoent _ -> []
+			)
+		|> List.concat
+		|> List.exists (fun vgpu ->
+			try
+				let path = Printf.sprintf "%s/pf" vgpu in
+				let pf = xs.Xs.read path in
+				pf = physical_function
+			with Xs_protocol.Enoent _ -> false
+		)
+	with Xs_protocol.Enoent _ -> false
+
+let call_gimtool args =
+	try
+		info "Initialising MxGPU PF: gimtool %s" (String.concat " " args);
+		ignore (Forkhelpers.execute_command_get_output !Xc_resources.gimtool args)
+	with _ ->
+		error "Failed to initialise MxGPU PF (gimtool %s)" (String.concat " " args);
```

Comment:
I would use `"(%s %s)" !Xc_resources.gimtool (String.concat " " args)` to have the real tool being called in the logs.